File: kalman.cpp
#include "opencv2/video/tracking.hpp"
#include "opencv2/highgui.hpp"

#include <iostream>

using namespace cv;
using namespace std;

void ontrack(int t,void *p)
{
    KalmanFilter * kf = (KalmanFilter*)p;
    double v = 1.0f / (t+1);
    cerr << v << endl;
    setIdentity(kf->processNoiseCov, Scalar::all(v*0.1));
    setIdentity(kf->measurementNoiseCov, Scalar::all(v)); 
    //kf->init(2,1);
}

void onmouse(int e,int x, int y, int d, void * ptr)
{
    Point3f * pt = (Point3f*)ptr;
    pt->x=float(x);
    pt->y=float(y);
    pt->z=float(d);
}

int main(int, char**)
{
    int kali=1000;
    Point3_<float> mouse;
    KalmanFilter KF(2, 2, 0);
    Mat measurement = Mat::zeros(2, 1, CV_32F);
    char code = (char)-1;
    namedWindow("Kalman",0);
    setMouseCallback("Kalman",onmouse,&mouse);
    createTrackbar("kali","Kalman",&kali,200000,ontrack,&KF);
    for(;;)
    {
        Mat img(500, 500, CV_8UC3);

        setIdentity(KF.measurementMatrix);
        setIdentity(KF.processNoiseCov, Scalar::all(.00005));
        setIdentity(KF.measurementNoiseCov, Scalar::all(0.002)); // this is inverse to the tracking speed
        setIdentity(KF.errorCovPost, Scalar::all(1));

        for(;;)
        {
            Mat prediction = KF.predict();
            if ( mouse.z == 1  )
            {
                float px = prediction.at<float>(0);
                float py = prediction.at<float>(1);
                circle( img, Point2f( px, py),  2, Scalar(20,200,20), 1, LINE_AA, 0); 
            }
            measurement.at<float>(0) = float(mouse.x);
            measurement.at<float>(1) = float(mouse.y);
            KF.correct(measurement);

            circle( img, Point2f( mouse.x, mouse.y),1, Scalar(180,180,180), 1, LINE_AA, 0); 
            imshow( "Kalman", img );

            code = (char)waitKey(10);
            if( code > 0 )
                break;
        }
        if( code == 27 || code == 'q' || code == 'Q' )
            break;
    }

    return 0;
}

File: demo/py/kalman.py
import cv2
import numpy as np

kalman = cv2.KalmanFilter(4,2)
measurement = np.array((2,1), np.float32)
pts_mouse = []
pts_track = []

def onmouse(b,x,y,s,nada):
    global pts_mouse, pts_track, measurement
    if s==1: # lbuttondown
        pts_mouse.append((x,y))
        measurement[0]=x
        measurement[1]=y

    if b==2: # rbuttonpress
        pts_mouse = []
        pts_track = []

def main():
    cv2.namedWindow("mouse",0)
    cv2.setMouseCallback("mouse", onmouse)

    kalman.measurementMatrix = np.array([[1,0,0,0],[0,1,0,0]],np.float32)
    kalman.transitionMatrix = np.array([[1,0,1,0],[0,1,0,1],[0,0,1,0],[0,0,0,1]],np.float32)
    kalman.processNoiseCov = np.array([[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0,0,1]],np.float32) * 0.003
    kalman.measurementNoiseCov = np.array([[1,0],[0,1]],np.float32) * 10
    kalman.errorCovPost = np.array([[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0,0,1]],np.float32) * 0.1

    while(True):
        prediction = kalman.predict()
        estimated = kalman.correct(measurement);

        #pts_track.append((prediction[0],prediction[1]))
        pts_track.append((estimated[0],estimated[1]))

        draw = np.zeros((400,400,3),np.uint8)
        if len(pts_mouse)>0:
            lp = pts_mouse[0]
            for p in pts_mouse:
                cv2.line(draw, lp, p, (200,0,0), 1)
                lp = p
            lp = pts_track[0]
            for p in pts_track:
                cv2.line(draw, lp, p, (0,0,200), 1)
                lp = p

        cv2.imshow("mouse", draw)
        if ((0xff & cv2.waitKey(1)) == 27):
            break
    cv2.destroyAllWindows()

main()
